File: modules/loofit-workout-core/ios/LoofitWorkoutCoreModule.swift
import ExpoModulesCore
import Foundation
import LoofitWorkoutCore

private enum LoofitWorkoutExpoAdapterError: LocalizedError {
  case configuration(String)
  case invalidCommand(String)

  var errorDescription: String? {
    switch self {
    case .configuration(let message), .invalidCommand(let message):
      return message
    }
  }
}

public final class LoofitWorkoutCoreModule: Module {
  private struct SurfaceContext: Sendable {
    let databaseDirectory: String
    let widgetsEnabled: Bool
  }

  private let contextLock = NSLock()
  private var lastSurfaceContext: SurfaceContext?
  private var timeZoneObserver: NSObjectProtocol?

  public func definition() -> ModuleDefinition {
    Name("LoofitWorkoutCore")

    Constant("widgetsConfigured") {
      LoofitWorkoutIntentEnvironment.widgetsAreConfigured()
    }

    Constant("widgetsDirectory") {
      LoofitWorkoutIntentEnvironment.databaseDirectory()
    }

    OnCreate {
      timeZoneObserver = NotificationCenter.default.addObserver(
        forName: .NSSystemTimeZoneDidChange,
        object: nil,
        queue: nil
      ) { [weak self] _ in
        guard let context = self?.surfaceContext() else { return }
        Task {
          _ = try? await LoofitWorkoutPipeline.reconcile(
            databaseDirectory: context.databaseDirectory,
            widgetsEnabled: context.widgetsEnabled
          )
        }
      }
    }

    OnDestroy {
      if let timeZoneObserver {
        NotificationCenter.default.removeObserver(timeZoneObserver)
      }
      timeZoneObserver = nil
    }

    AsyncFunction("executeWorkoutCommand") {
      (record: LoofitWorkoutCommandRecord, directory: String?, widgetsEnabled: Bool) async throws
        -> LoofitWorkoutCommandResultRecord in
      let databaseDirectory = try resolveAndRemember(directory, widgetsEnabled: widgetsEnabled)
      let result = try await LoofitWorkoutPipeline.execute(
        try record.command(),
        databaseDirectory: databaseDirectory,
        widgetsEnabled: widgetsEnabled
      )
      return LoofitWorkoutCommandResultRecord(result)
    }

    AsyncFunction("reconcileWorkoutSurfaces") {
      (directory: String?, widgetsEnabled: Bool) async throws -> LoofitWorkoutCommandResultRecord in
      let databaseDirectory = try resolveAndRemember(directory, widgetsEnabled: widgetsEnabled)
      let result = try await LoofitWorkoutPipeline.reconcile(
        databaseDirectory: databaseDirectory,
        widgetsEnabled: widgetsEnabled
      )
      return LoofitWorkoutCommandResultRecord(result)
    }

    AsyncFunction("updateWidgetThemeSnapshot") {
      (record: LoofitWidgetThemeRecord, directory: String?, widgetsEnabled: Bool) async throws
        -> LoofitWorkoutCommandResultRecord in
      let databaseDirectory = try resolveAndRemember(directory, widgetsEnabled: widgetsEnabled)
      let result = try await LoofitWorkoutPipeline.updateTheme(
        record.theme,
        databaseDirectory: databaseDirectory,
        widgetsEnabled: widgetsEnabled
      )
      return LoofitWorkoutCommandResultRecord(result)
    }
  }

  private func resolveAndRemember(_ directory: String?, widgetsEnabled: Bool) throws -> String {
    let widgetsConfigured = LoofitWorkoutIntentEnvironment.widgetsAreConfigured()
    guard widgetsConfigured == widgetsEnabled else {
      throw LoofitWorkoutExpoAdapterError.configuration(
        "The requested widget mode does not match the installed iOS build configuration."
      )
    }
    let databaseDirectory = try resolvedDirectory(
      directory,
      widgetsConfigured: widgetsConfigured
    )
    contextLock.lock()
    lastSurfaceContext = .init(
      databaseDirectory: databaseDirectory,
      widgetsEnabled: widgetsConfigured
    )
    contextLock.unlock()
    return databaseDirectory
  }

  private func surfaceContext() -> SurfaceContext? {
    contextLock.lock()
    defer { contextLock.unlock() }
    return lastSurfaceContext
  }

  private func resolvedDirectory(
    _ directory: String?,
    widgetsConfigured: Bool
  ) throws -> String {
    if widgetsConfigured {
      let appGroupDirectory = try LoofitWorkoutIntentEnvironment.requireDatabaseDirectory()
      if let value = directory?.trimmingCharacters(in: .whitespacesAndNewlines),
        !value.isEmpty,
        LoofitWorkoutPaths.normalizeDatabaseDirectory(value) != appGroupDirectory
      {
        throw LoofitWorkoutExpoAdapterError.configuration(
          "The requested database directory does not match the configured App Group directory."
        )
      }
      return appGroupDirectory
    }

    guard let value = directory?.trimmingCharacters(in: .whitespacesAndNewlines),
          !value.isEmpty else {
      return try LoofitWorkoutPaths.defaultDatabaseDirectory()
    }
    return LoofitWorkoutPaths.normalizeDatabaseDirectory(value)
  }
}

struct LoofitWorkoutCommandRecord: Record {
  @Field var type: String = ""
  @Field var routineDayId: Int64?
  @Field var expectedSessionId: Int64?
  @Field var bodyPartIds: [Int64] = []
  @Field var label: String?

  func command() throws -> LoofitWorkoutCommand {
    switch type {
    case "startNext":
      return .startNext
    case "startRoutine":
      guard let routineDayId else {
        throw LoofitWorkoutExpoAdapterError.invalidCommand("startRoutine requires routineDayId")
      }
      return .startRoutine(routineDayId: routineDayId)
    case "startFree":
      return .startFree(bodyPartIds: bodyPartIds, label: label)
    case "changeRoutine":
      guard let expectedSessionId, let routineDayId else {
        throw LoofitWorkoutExpoAdapterError.invalidCommand(
          "changeRoutine requires expectedSessionId and routineDayId"
        )
      }
      return .changeRoutine(
        expectedSessionId: expectedSessionId,
        routineDayId: routineDayId
      )
    case "changeFree":
      guard let expectedSessionId else {
        throw LoofitWorkoutExpoAdapterError.invalidCommand("changeFree requires expectedSessionId")
      }
      return .changeFree(
        expectedSessionId: expectedSessionId,
        bodyPartIds: bodyPartIds
      )
    case "complete":
      guard let expectedSessionId else {
        throw LoofitWorkoutExpoAdapterError.invalidCommand("complete requires expectedSessionId")
      }
      return .complete(expectedSessionId: expectedSessionId)
    case "cancel":
      guard let expectedSessionId else {
        throw LoofitWorkoutExpoAdapterError.invalidCommand("cancel requires expectedSessionId")
      }
      return .cancel(expectedSessionId: expectedSessionId)
    default:
      throw LoofitWorkoutExpoAdapterError.invalidCommand("Unsupported workout command: \(type)")
    }
  }
}

struct LoofitWidgetThemeRecord: Record {
  @Field var brandName: String = ""
  @Field var accent: String = ""
  @Field var accentText: String = ""
  @Field var background: String = ""
  @Field var labelColor: String = ""
  @Field var brandColor: String = ""
  @Field var titleColor: String = ""
  @Field var detailColor: String = ""
  @Field var secondaryButtonBackground: String = ""
  @Field var secondaryButtonText: String = ""
  @Field var heatmapBackground: String = ""
  @Field var heatmapTitleColor: String = ""
  @Field var heatmapBrandColor: String = ""
  @Field var heatmapFooterValueColor: String = ""
  @Field var heatmapWeekdayLabelColor: String = ""
  @Field var heatmapWeekendLabelColor: String = ""
  @Field var heatmapDayLabelColor: String = ""
  @Field var heatmapBaseColor: String = ""
  @Field var heatmapEmptyColor: String = ""
  @Field var heatmapGapColor: String = ""
  @Field var todayIndicatorColor: String = ""

  var theme: LoofitWidgetTheme {
    LoofitWidgetTheme(dictionary: [
      "brandName": brandName,
      "accent": accent,
      "accentText": accentText,
      "background": background,
      "labelColor": labelColor,
      "brandColor": brandColor,
      "titleColor": titleColor,
      "detailColor": detailColor,
      "secondaryButtonBackground": secondaryButtonBackground,
      "secondaryButtonText": secondaryButtonText,
      "heatmapBackground": heatmapBackground,
      "heatmapTitleColor": heatmapTitleColor,
      "heatmapBrandColor": heatmapBrandColor,
      "heatmapFooterValueColor": heatmapFooterValueColor,
      "heatmapWeekdayLabelColor": heatmapWeekdayLabelColor,
      "heatmapWeekendLabelColor": heatmapWeekendLabelColor,
      "heatmapDayLabelColor": heatmapDayLabelColor,
      "heatmapBaseColor": heatmapBaseColor,
      "heatmapEmptyColor": heatmapEmptyColor,
      "heatmapGapColor": heatmapGapColor,
      "todayIndicatorColor": todayIndicatorColor,
    ])
  }
}

struct LoofitWorkoutCommandResultRecord: Record {
  @Field var status: String = LoofitWorkoutCommandStatus.rejected.rawValue
  @Field var sessionId: Int64?
  @Field var desiredRevision: Int64 = 0
  @Field var publishedRevision: Int64 = 0
  @Field var publicationStatus: String = LoofitWorkoutPublicationStatus.skipped.rawValue
  @Field var publicationError: String?

  init() {}

  init(_ result: LoofitWorkoutCommandResult) {
    status = result.status.rawValue
    sessionId = result.sessionId
    desiredRevision = result.desiredRevision
    publishedRevision = result.publishedRevision
    publicationStatus = result.publicationStatus.rawValue
    publicationError = result.publicationError
  }
}

File: modules/loofit-workout-core/android/src/main/java/com/loofit/workoutcore/LoofitWorkoutCoreModule.kt
package com.loofit.workoutcore

import android.content.Context
import com.google.android.play.core.appupdate.AppUpdateManagerFactory
import com.google.android.play.core.install.model.UpdateAvailability
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class LoofitWorkoutCoreModule : Module() {
  private val externalCommandListener: (LoofitWorkoutCommandResult) -> Unit = { result ->
    sendEvent(EXTERNAL_COMMAND_EVENT, result.dictionary())
  }

  override fun definition() = ModuleDefinition {
    Name("LoofitWorkoutCore")

    Constant("widgetsConfigured") { true }
    Constant("widgetsDirectory") { null as String? }
    Events(EXTERNAL_COMMAND_EVENT)

    OnStartObserving(EXTERNAL_COMMAND_EVENT) {
      LoofitWorkoutCommandEvents.addListener(externalCommandListener)
    }

    OnStopObserving(EXTERNAL_COMMAND_EVENT) {
      LoofitWorkoutCommandEvents.removeListener(externalCommandListener)
    }

    AsyncFunction("executeWorkoutCommand") {
        record: Map<String, Any?>,
        _: String?,
        widgetsEnabled: Boolean,
      ->
      checkWidgetMode(widgetsEnabled)
      LoofitWorkoutPipeline.execute(context(), record.command(), widgetsEnabled).dictionary()
    }

    AsyncFunction("reconcileWorkoutSurfaces") {
        _: String?,
        widgetsEnabled: Boolean,
      ->
      checkWidgetMode(widgetsEnabled)
      LoofitWorkoutPipeline.reconcile(context(), widgetsEnabled).dictionary()
    }

    AsyncFunction("updateWidgetThemeSnapshot") {
        record: Map<String, Any?>,
        _: String?,
        widgetsEnabled: Boolean,
      ->
      checkWidgetMode(widgetsEnabled)
      LoofitWorkoutPipeline.updateTheme(
        context(),
        LoofitWidgetTheme.fromMap(record),
        widgetsEnabled,
      ).dictionary()
    }

    AsyncFunction("getAndroidUpdateInfo") { promise: Promise ->
      AppUpdateManagerFactory.create(context()).appUpdateInfo
        .addOnSuccessListener { info ->
          promise.resolve(
            mapOf(
              "updateAvailable" to
                (info.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE),
              "availableVersionCode" to info.availableVersionCode(),
            ),
          )
        }
        .addOnFailureListener { error ->
          promise.reject(
            "ERR_LOOFIT_PLAY_UPDATE",
            error.message ?: "Google Play update lookup failed.",
            error,
          )
        }
    }
  }

  private fun context(): Context = appContext.reactContext?.applicationContext
    ?: throw LoofitWorkoutCoreException("Android application context is unavailable.")

  private fun checkWidgetMode(widgetsEnabled: Boolean) {
    if (!widgetsEnabled) {
      throw LoofitWorkoutCoreException(
        "The requested widget mode does not match the installed Android build configuration.",
      )
    }
  }

  companion object {
    const val EXTERNAL_COMMAND_EVENT = "onExternalWorkoutCommand"
  }
}

private class LoofitWorkoutCoreException(message: String) : CodedException(message)

private fun Map<String, Any?>.command(): LoofitWorkoutCommand = when (this["type"] as? String) {
  "startNext" -> LoofitWorkoutCommand.StartNext
  "startRoutine" -> LoofitWorkoutCommand.StartRoutine(requiredLong("routineDayId"))
  "changeParts" -> LoofitWorkoutCommand.ChangeParts(
    expectedSessionId = requiredLong("expectedSessionId"),
    bodyPartIds = (this["bodyPartIds"] as? List<*>)
      .orEmpty()
      .map { (it as? Number)?.toLong() ?: throw LoofitWorkoutCoreException("Invalid bodyPartId.") },
    updateRoutine = this["updateRoutine"] as? Boolean ?: false,
  )
  "complete" -> LoofitWorkoutCommand.Complete(requiredLong("expectedSessionId"))
  "cancel" -> LoofitWorkoutCommand.Cancel(requiredLong("expectedSessionId"))
  else -> throw LoofitWorkoutCoreException("Unsupported workout command type.")
}

private fun Map<String, Any?>.requiredLong(key: String): Long =
  (this[key] as? Number)?.toLong()
    ?: throw LoofitWorkoutCoreException("Workout command requires $key.")
